File: src/section-parser.ts
import { Section } from "./definition";

class Parser {
	lines: string[];
	line: string;
	inSection: boolean;
	constructor(input: string) {
		this.lines = input.slice(1).split(/[\r\n]+/g);
		this.line = "";
		this.inSection = false;
	}

	parse() {
		const out: Section[] = [];
		const len = this.lines.length;

		for (let i = 0; i < len; i++) {
			const line = this.lines[i];
			if (!line || line.match(/^\s*[;#]/)) {
				continue;
			}
			this.line = line.trim();

			this.parseLine(out);
		}

		return out;
	}

	parseLine(out: Section[]) {
		if (this.inSection) {
			const header = out[out.length - 1];

			return this.findClose(header);
		}

		const sectionHeader = this.parseSecHeader();
		out.push(sectionHeader);
	}

	parseSecHeader(): Section {
		const match = /^\[([A-Z]+)](\d+)$/.exec(this.line);
		if (!match) {
			throw new SyntaxError("no header");
		}

		this.inSection = true;
		const name = match[1];
		const order = Number(match[2]);

		return { name, order, lines: [] };
	}

	findClose(header: Section) {
		const closing = `[/${header.name}]`;
		if (this.line === closing) {
			this.inSection = false;
		} else {
			header.lines.push(this.line);
		}
	}
}

export function parseSection(input: string) {
	return new Parser(input).parse();
}

export default Parser;

File: section_parser.go
package main

import (
	"log"
	"regexp"
	"strconv"
	"strings"
)

type Parser struct {
	lines     []string
	line      string
	inSection bool
}

type Section struct {
	name  string
	order int64
	lines []string
}

func ParseString(input string) []*Section {
	parser := &Parser{
		lines: regexp.MustCompile("[\r\n]+").Split(input, -1),
	}

	return parser.Parse()
}

var emptyLineTest = regexp.MustCompile(`^\s*[;#]`)

func (p *Parser) Parse() []*Section {
	out := make([]*Section, 0)
	length := len(p.lines)
	for i := 0; i < length; i++ {
		line := p.lines[i]
		if line == "" || emptyLineTest.MatchString(line) {
			continue
		}
		p.line = strings.TrimSpace(line)

		out = p.ParseLine(out)
	}

	return out
}

func (p *Parser) ParseLine(out []*Section) []*Section {
	if p.inSection {
		header := out[len(out)-1]
		p.FindClose(header)
		return out
	}

	sectionHeader := p.ParseSecHeader()
	out = append(out, sectionHeader)
	return out
}

var sectionHeaderTest = regexp.MustCompile(`^\[([A-Z_]+)](\d+)$`)

func (p *Parser) ParseSecHeader() *Section {
	if !sectionHeaderTest.MatchString(p.line) {
		log.Fatalln("No Header:", p.line, "len", len(p.line))
	}
	section := &Section{}

	p.inSection = true
	section.lines = make([]string, 0)
	match := sectionHeaderTest.FindStringSubmatch(p.line)
	section.name = match[1]
	section.order, _ = strconv.ParseInt(match[2], 10, 64)
	return section
}

func (p *Parser) FindClose(header *Section) {
	closing := "[/" + header.name + "]"
	if p.line == closing {
		p.inSection = false
	} else {
		header.lines = append(header.lines, p.line)
	}
}
